### PR TITLE
Move jailing code to core #1

### PR DESCRIPTION
**Pull request by IljaN**
## Description
- Read-only jailing of guest users is now provided by core
- Read-only groups can be configured in oc_appconfig under 'core' => 'read_only_groups' (coma seperated)

## Related Issue
#https://github.com/owncloud/guests/issues/1

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
Looks good so far.

- move the wrapper registration after the presetup hook firing
- use pipe separated group names (I thought we had this in sharing code already) because LDAP groups can have commas...
- only apply read only wrapper for logged in user ? note that setup FS might be called when mounting shares from other users in which case the wrapper might be superfluous (to be checked with a debugger)
- verify that fetching group memberships doesn't cause additional SQL queries. Normally the group manager already cached that info so we might be fine already

**Comment by IljaN** [comment]
@PVince81 I am not sure if I understand. The code is already applying the wrapper only for logged in users. 

> I thought we had this in sharing code already
Do you imply that there is some code I can reuse?

**Comment by IljaN** [comment]
@PVince81 The Problem of this code is that when a external storage is Mounted for all users, and the guest user accesses it he suddenly gains write permission :-/

**Comment by PVince81** [comment]
> The code is already applying the wrapper only for logged in users.

I meant "the currently logged in user as opposed to when `setupFS` is called for sharing users which aren't necessarily the current user". Need to test whether this actually happens.

**Comment by PVince81** [comment]
> Do you imply that there is some code I can reuse?

I remembered it wrong. Actually you need to use JSON instead.
Grep for the key "shareapi_exclude_groups_list" which match the sharing setting "Exclude groups from sharing" in the admin page.

>  The Problem of this code is that when a external storage is Mounted for all users, and the guest user accesses it he suddenly gains write permission

Good catch. I think the original code didn't care about such use cases (@butonic), but I think we should change it to also apply the wrapper on external storages. You might need to register the wrapper in a different location in the code, maybe close to the encryption or checksum wrapper as they also wrap every possible storage except sharing.

**Comment by IljaN** [comment]
Yeah the problem is that the SFTP stuff is done after setupFs is executed

**Comment by PVince81** [comment]
The "oc_availability" storage also works for external storages, so I think you should register the jail one just after that one.

File: lib/private/legacy/util.php [hunk 1]
```diff
@@ -208,9 +210,48 @@ public static function setupFS($user = '') {
...
 		OC_Hook::emit('OC_Filesystem', 'preSetup', ['user' => $user]);
 		\OC\Files\Filesystem::logWarningWhenAddingStorageWrapper(true);
 
+		// Make users storage readonly if he is in a read only group
+
+		$readOnlyGroups = json_decode(\OC::$server->getConfig()->getAppValue(
+			'core',
+			'read_only_groups',
+			[]
+		));
```
Comment (PVince81) [inline lib/private/legacy/util.php]: `json_decode($something, true)`, else it will create a dummy class

Comment (PVince81) [inline lib/private/legacy/util.php]: and I can imagine some people trying to use frontend side injection to make that class do wild things 😄

Comment (IljaN) [inline lib/private/legacy/util.php]: > and I can imagine some people trying to use frontend side injection to make that class do wild  >things :smile:

I will include some sanitizing at the write site?

Comment (PVince81) [inline lib/private/legacy/util.php]: > I will include some sanitizing at the write site?

Yes, that too

**Comment by IljaN** [comment]
@PVince81 should @felixheidecke integrate the required changes for the frontend here or in a seperate PR? (Configuration for read_only_groups)

**Comment by PVince81** [comment]
@IljaN up to you guys

**Comment by felixheidecke** [comment]
I prefer working on the same branch if the team consists of no more than two peeps :wink:

**Comment by IljaN** [comment]
@PVince81 This implementation prevents the creation of the user directory after first login. Any ideas?

**Comment by IljaN** [comment]
Rebase Fail: continue here: https://github.com/owncloud/core/pull/27286

**Pull request closed without merging**